File: app/src/main/java/com/hellmund/primetime/ui/introduction/IntroductionActivity.java
package com.hellmund.primetime.ui.introduction;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.AsyncTaskLoader;
import android.support.v4.content.Loader;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.widget.GridView;

import com.hellmund.primetime.R;
import com.hellmund.primetime.ui.SelectGenreActivity;
import com.hellmund.primetime.utils.DeviceUtils;
import com.hellmund.primetime.utils.DownloadManager;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class IntroductionActivity extends FragmentActivity {

    @BindView(R.id.grid_view) GridView mGridView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_introduction);

        getWindow().setBackgroundDrawable(null);
        ButterKnife.bind(this);

        if (DeviceUtils.isLandscapeMode(this)) {
            mGridView.setNumColumns(4);
        }

        downloadPopularMovies();
    }

    private void downloadPopularMovies() {
        getSupportLoaderManager().initLoader(0, null,
                new LoaderManager.LoaderCallbacks<ArrayList<String>>() {
            @NonNull
            @Override
            public Loader<ArrayList<String>> onCreateLoader(int id, Bundle args) {
                return new PopularMoviesLoader(IntroductionActivity.this);
            }

            @Override
            public void onLoadFinished(Loader<ArrayList<String>> loader, ArrayList<String> data) {
                displayResults(data);
            }

            @Override
            public void onLoaderReset(Loader<ArrayList<String>> loader) {}
        });
    }

    private void displayResults(ArrayList<String> results) {
        PostersAdapter adapter = new PostersAdapter(this, results);
        mGridView.setAdapter(adapter);
        mGridView.setVisibility(View.VISIBLE);

        mGridView.setOnTouchListener((v, event) -> event.getAction() == MotionEvent.ACTION_MOVE);

        Animation fadeIn = new AlphaAnimation(0.0f, 1.0f);
        fadeIn.setDuration(700);
        mGridView.setAnimation(fadeIn);
    }

    @OnClick(R.id.introduction_btn)
    public void openGenresSelection() {
        Intent intent = new Intent(this, SelectGenreActivity.class);
        startActivity(intent);
    }

    private static class PopularMoviesLoader extends AsyncTaskLoader<ArrayList<String>> {

        private ArrayList<String> mResults;

        PopularMoviesLoader(Context context) {
            super(context);
        }

        @Override
        protected void onStartLoading() {
            if (mResults != null) {
                deliverResult(mResults);
            } else {
                forceLoad();
            }
        }

        @Override
        public ArrayList<String> loadInBackground() {
            return DownloadManager.downloadMostPopularMoviePosters();
        }

        @Override
        public void deliverResult(ArrayList<String> results) {
            mResults = results;
            super.deliverResult(results);
        }

    }

}

File: app/src/main/java/com/hellmund/primetime/ui/introduction/IntroductionActivity.kt
package com.hellmund.primetime.ui.introduction

import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.animation.OvershootInterpolator
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.GridLayoutManager
import com.hellmund.primetime.R
import com.hellmund.primetime.di.injector
import com.hellmund.primetime.di.lazyViewModel
import com.hellmund.primetime.ui.selectgenres.SelectGenresActivity
import com.hellmund.primetime.utils.ImageLoader
import com.hellmund.primetime.utils.observe
import kotlinx.android.synthetic.main.activity_introduction.gridView
import kotlinx.android.synthetic.main.activity_introduction.introductionButton
import java.util.Timer
import javax.inject.Inject
import javax.inject.Provider
import kotlin.concurrent.schedule


class IntroductionActivity : AppCompatActivity() {

    @Inject
    lateinit var imageLoader: ImageLoader

    @Inject
    lateinit var viewModelProvider: Provider<IntroductionViewModel>

    private val viewModel: IntroductionViewModel by lazyViewModel { viewModelProvider }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_introduction)
        injector.inject(this)

        startButtonAnimation()

        introductionButton.setOnClickListener { openGenresSelection() }
        viewModel.posterUrls.observe(this, this::displayResults)
    }

    // TODO: Find an easier way to do this
    private fun startButtonAnimation() {
        Timer().schedule(500) { hideButton() }
        Timer().schedule(1_800) { slideInButton() }
    }

    private fun hideButton() {
        val buttonHeight = introductionButton.height
        val buttonMargin = resources.getDimensionPixelOffset(R.dimen.default_space)
        val verticalButtonTranslation = (buttonHeight + buttonMargin).toFloat()

        val handler = Handler(Looper.getMainLooper())
        handler.post {
            introductionButton.animate()
                    .setDuration(400L)
                    .translationYBy(verticalButtonTranslation)
                    .setInterpolator(OvershootInterpolator(1f))
                    .start()
        }
    }

    private fun slideInButton() {
        val buttonHeight = introductionButton.height
        val buttonMargin = resources.getDimensionPixelOffset(R.dimen.default_space)
        val verticalButtonTranslation = (buttonHeight + buttonMargin).toFloat()

        val handler = Handler(Looper.getMainLooper())
        handler.post {
            introductionButton.animate()
                    .alphaBy(1f)
                    .translationYBy(verticalButtonTranslation * (-1))
                    .setInterpolator(OvershootInterpolator(1f))
                    .start()
        }
    }

    private fun displayResults(results: List<String>) {
        val adapter = PostersAdapter(imageLoader, results)
        gridView.itemAnimator = DefaultItemAnimator()
        gridView.adapter = adapter
        gridView.isEnabled = false

        val columns = if (isLandscapeMode) 4 else 3
        gridView.layoutManager = GridLayoutManager(this, columns)
    }

    private fun openGenresSelection() {
        val intent = SelectGenresActivity.newIntent(this)
        startActivity(intent)
    }

    private val Context.isLandscapeMode: Boolean
        get() = resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE

}
